src/Hive.Versioning/Parsing/ErrorMessages.cs

```diff
@@ -0,0 +1,883 @@
+                    comparer = new(comparer.CompareTo, newCompare);
+                }
+
+                if (GetDirection(comparer1) == GetDirection(comparer2))
+                {
+                    // if they have the same direction, we should flip one of them
+                    // which one we should flip depends on their direction
+                    if (GetDirection(comparer1) == VersionRange.ComparisonType.Less)
+                        FlipDirection(ref comparer1); // if they both point down, flip lower
+                    else if (GetDirection(comparer1) == VersionRange.ComparisonType.Greater)
+                        FlipDirection(ref comparer2); // if they both point up, flip upper
+                }
+
+                if (GetDirection(comparer1) == VersionRange.ComparisonType.Less
+                    || GetDirection(comparer2) == VersionRange.ComparisonType.Greater)
+                {
+                    // if at this point they are pointing the wrong direction , they should both be swapped.
+                    FlipDirection(ref comparer1);
+                    FlipDirection(ref comparer2);
+                }
+
+                // the lower bound must match the upper bound
+                if (!comparer1.Matches(comparer2))
+                {
+                    // if it doesn't, then we should add an equals
+                    comparer1 = new(comparer1.CompareTo, comparer1.Type | VersionRange.ComparisonType.ExactEqual);
+                }
+                // the upper bound must match the lower bound
+                if (!comparer2.Matches(comparer1))
+                {
+                    // if it doesn't, then we should add an equals
+                    comparer2 = new(comparer2.CompareTo, comparer2.Type | VersionRange.ComparisonType.ExactEqual);
+                }
+            }
+
+            // now we should have a pretty much valid pair of comparers
+            var sb = new StringBuilder();
+            if (!isSingle) sb = comparer1.ToString(sb).Append(' ');
+            sb = comparer2.ToString(sb);
+
+            return new(SR.Range_BoundedRegionNotClosed, Suggestion: sb.ToString(), Span: SpanFromReport(reports[start]));
+        }
+        #endregion
+
+        private static GeneratedMessage ExtraInputMesage<T>(in MessageInfo msgs, ActionErrorReport<T> report)
+            where T : struct
+            => new(SR.ExtraInputAtEnd, Suggestion: "", Span: SpanFromReport(report));
+
+        private static (int Start, int Length) ScanForErrorRange<T>(IReadOnlyList<ActionErrorReport<T>> reports, Func<T, bool> checkIsError)
+            where T : struct
+        {
+            var end = reports.Count - 1;
+            var position = end;
+
+            while (position >= 0 && checkIsError(reports[position].Action))
+                position--;
+
+            return (position + 1, end - position);
+        }
+
+        private static bool VersionIsError(VersionParseAction action)
```

Comment:
Boy, wouldn't it be cool if enums could implement interfaces... Would make this much nicer, for sure...

Reply:
Eh, part of me prefers doing it like this, as annoying as it can be.